### PR TITLE
Add category management to the bookstore app

### PR DESCRIPTION
**Pull request by chertiav**
- introduced features for managing categories.
- implemented a new CategoryService and respective repository and controller classes.
- enhanced error handling with a common API error response class.

### CODE REVIEW

File: src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java [hunk 1]
```diff
@@ -34,4 +35,5 @@ public class CreateBookRequestDto {
 
     private String description;
     private String coverImage;
+    private List<Long> categories;
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java:40]: do we need to validate it?

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -29,6 +33,14 @@ public Page<BookDto> findAll(Pageable pageable) {
                 .map(bookMapper::toDto);
     }
 
+    @Override
+    public Page<BookDtoWithoutCategoryIds> findAllByCategoryId(Long categoryId, Pageable pageable) {
+        Category category = new Category(categoryId);
+        Set<Category> categorySet = Collections.singleton(category);
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -29,6 +33,14 @@ public Page<BookDto> findAll(Pageable pageable) {
...
+        return bookRepository.findAllByCategories(categorySet, pageable)
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAllByCategoriesId(categoryId, pageable)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java (hunk 1 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java:40]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java (hunk 2 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java (hunk 3 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/service/impl/BookServiceImpl.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Categories management", description = "Endpoints for managing categories")
+@RestController
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+    private static final String RESPONSE_CODE_OK = "200";
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/controller/CategoryController.java]: If you have similar constants in each controller, it's better to replace it to util class (for example)

File: src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java [hunk 5]
```diff
@@ -34,4 +36,7 @@ public class CreateBookRequestDto {
 
     private String description;
     private String coverImage;
+    @NotNull(message = "Categories cannot be null")
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
two comments

File: src/main/java/com/achdev/onlinebookstoreapp/controller/CategoryController.java (hunk 4 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/controller/CategoryController.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java (hunk 5 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/dto/book/CreateBookRequestDto.java]: Ok

**Review by liliia-ponomarenko — APPROVED** [review]
Good job!

**Pull request merged**